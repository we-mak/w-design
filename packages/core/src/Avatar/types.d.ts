import { GlobProps, SizeProps } from "../typings";

export interface AvatarProps extends SizeProps, GlobProps {
  size?: "xs" | "sm" | "md" | "lg" | "xl" | "xxl";
  // User online status
  presence?: "online" | "offline" | "busy";
  avatarUrl?: string;
  // Initial content of avatar if without image
  dataInitial?: string;
  alt?: string;
}
