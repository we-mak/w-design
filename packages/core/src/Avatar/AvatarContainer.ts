import styled from "styled-components";
import { space } from "styled-system";
import { AvatarProps } from "./index";

const getSize = (props: AvatarProps) => {
  switch (props.size) {
    case "xs":
      return `
      height: 1.4rem;
      width: 1.4rem;
      `;
    case "sm":
      return `
      height: 2rem;
      width: 2rem;
      `;
    case "md":
      return `
      height: 2.8rem;
      width: 2.8rem;
      `;
    case "lg":
      return `
      height: 3.2rem;
      width: 3.2rem;
      `;
    case "xl":
      return `
      height: 4rem;
      width: 4rem;
      `;
    case "xxl":
      return `
      height: 6rem;
      width: 6rem;
      `;
    default:
      return `
    height: 1.8rem;
    width: 1.8rem;
    `;
  }
};

export const AvatarContainer = styled.figure<AvatarProps>`
  border-radius: 50%;
  color: rgba(255, 255, 255, 0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  ${getSize};
  ${space};
`;
