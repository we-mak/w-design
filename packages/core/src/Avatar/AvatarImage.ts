import styled from "styled-components";
import Image from "../Image";

export const AvatarImage = styled(Image)`
  border-radius: 50%;
  height: 100%;
  width: 100%;
  position: relative;
  z-index: 1;
`;
