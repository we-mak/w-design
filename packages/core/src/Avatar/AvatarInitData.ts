import styled from "styled-components";
import { AvatarProps } from "./index";

const getTextSize = (props: AvatarProps) => {
  switch (props.size) {
    case "xs":
      return `font-size: .6rem;`;
    case "sm":
      return `font-size: .9rem;`;
    case "md":
      return `font-size: 1rem;`;
    case "lg":
      return `font-size: 1.8rem;`;
    case "xl":
      return `font-size: 2rem;`;
    case "xxl":
      return `font-size: 2.8rem;`;
    default:
      return `font-size: 1.4rem;`;
  }
};

export const AvatarInitData = styled.span<AvatarProps>`
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  ${getTextSize};
`;
