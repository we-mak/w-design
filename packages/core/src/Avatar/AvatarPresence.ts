import styled from "styled-components";
import { AvatarProps } from "./index";

const getPresenceStyle = (props: AvatarProps) => {
  switch (props.size) {
    case "xs":
      return `
    height: .4rem;
    width: .4rem;
    right: 0;
    border: 0.075rem solid #fff;
    `;
    case "sm":
      return `
    height: .5rem;
    width: .5rem;
    right: 0;
    border: 0.075rem solid #fff;
    `;
    case "md":
      return `
    height: .7rem;
    width: .7rem;
    right: 0;
    border: 0.075rem solid #fff;
    `;
    case "lg":
      return `
    height: .7rem;
    width: .7rem;
    right: .1rem;
    border: 0.075rem solid #fff;
    `;

    case "xl":
      return `
    height: .8rem;
    width: .8rem;
    right: .3rem;
    border: 0.1rem solid #fff;
    `;
    case "xxl":
      return `
    height: .9rem;
    width: .9rem;
    right: .9rem;
    border: 0.1rem solid #fff;
    `;
    default:
      return `
    height: .7rem;
    width: .7rem;
    right: .7rem;
    border: 0.1rem solid #fff;
  `;
  }
};

const setPresenceColor = (props: AvatarProps) => {
  const {
    presence,
    theme: { colors }
  } = props;

  if (presence === "online") return colors["G30"];
  if (presence === "offline") return colors["N20"];
  if (presence === "busy") return colors["R30"];

  return colors["G30"];
};

export const AvatarPresence = styled.span<AvatarProps>`
  border-radius: 50%;
  bottom: 0.1rem;
  pointer-events: none;
  position: absolute;
  z-index: 30;
  background-color: ${setPresenceColor};
  ${getPresenceStyle};
`;
