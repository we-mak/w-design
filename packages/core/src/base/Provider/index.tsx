import * as React from "react";
import { ThemeProvider } from "styled-components";
import defaultTheme from "../../common/theme";
import { GlobalStyle } from "./GlobalStyle";
import "./normalize.css";

interface ThemeProps extends React.HTMLProps<HTMLDivElement> {
  theme?: object | ((theme: object) => object);
}

function Provider(p: ThemeProps) {
  const { theme = {}, children } = p;

  return (
    <ThemeProvider theme={{ ...defaultTheme, ...theme }}>
      <>
        <GlobalStyle />
        {children}
      </>
    </ThemeProvider>
  );
}

Provider.displayName = "WProvider";

export default Provider;
