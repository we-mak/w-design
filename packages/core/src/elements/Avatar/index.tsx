import * as React from "react";
import {
  AvatarWrapper,
  AvatarDataInit,
  AvatarImage,
  PresenceMark
} from "./Styled";
import { firstLetter, randomProperty, filterObject } from "@w-design/helpers";
import { AvatarProps } from "./types";
import { colors } from "../../common/colors";

const Avatar = ({
  size = "md",
  avatarUrl,
  dataInitial = "We mak",
  alt,
  presence,
  ...rest
}: AvatarProps) => {
  const [backgroundColor, setBackgroundColor] = React.useState("");

  React.useEffect(() => {
    if (!avatarUrl) {
      setBackgroundColor(
        randomProperty(filterObject(colors, ["WHITE", "BLACK", "N1"]))
      );
    }
  }, []);

  return (
    <AvatarWrapper style={{ backgroundColor }} size={size} {...rest}>
      {dataInitial && (
        <AvatarDataInit size={size}>{firstLetter(dataInitial)}</AvatarDataInit>
      )}
      {avatarUrl && <AvatarImage data-src={avatarUrl} alt={alt} />}
      {presence && <PresenceMark presence={presence} size={size} />}
    </AvatarWrapper>
  );
};

export default React.memo(Avatar);
export { AvatarProps };
