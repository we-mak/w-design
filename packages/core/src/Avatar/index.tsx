import * as React from "react";
import { firstLetter, randomProperty, filterObject } from "@w-design/helpers";
import { colors } from "../common/colors";
import { GlobProps } from "../common/props";
import { AvatarContainer } from "./AvatarContainer";
import { AvatarPresence } from "./AvatarPresence";
import { AvatarInitData } from "./AvatarInitData";
import { AvatarImage } from "./AvatarImage";

export interface AvatarProps extends GlobProps {
  size?: "xs" | "sm" | "md" | "lg" | "xl" | "xxl";
  // User online status
  presence?: "online" | "offline" | "busy";
  avatarUrl?: string;
  // Initial content of avatar if without image
  dataInitial?: string;
  alt?: string;
  crossOrigin?: React.ImgHTMLAttributes<any>["crossOrigin"];
}

const Avatar = ({
  size = "md",
  avatarUrl,
  dataInitial = "We mak",
  alt,
  crossOrigin,
  presence,
}: AvatarProps) => {
  const [backgroundColor, setBackgroundColor] = React.useState("white");

  React.useEffect(() => {
    if (!avatarUrl) {
      setBackgroundColor(
        randomProperty(filterObject(colors, ["WHITE", "BLACK", "N1"]))
      );
    }
  }, []);

  return (
    <AvatarContainer style={{ backgroundColor }} size={size}>
      {dataInitial && (
        <AvatarInitData size={size}>{firstLetter(dataInitial)}</AvatarInitData>
      )}
      {avatarUrl && (
        <AvatarImage data-src={avatarUrl} alt={alt} crossOrigin={crossOrigin} />
      )}
      {presence && <AvatarPresence presence={presence} size={size} />}
    </AvatarContainer>
  );
};

export default React.memo(Avatar);
