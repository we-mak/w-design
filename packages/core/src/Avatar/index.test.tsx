import * as React from "react";
import { render, cleanup } from "@testing-library/react";
import Provider from "../Provider";
import Avatar from "./";

afterEach(cleanup);

describe("<Avatar/>", () => {
  // Render test
  it("should renders correctly", () => {
    const { container } = render(
      <Provider>
        <Avatar
          size="md"
          alt="test"
          presence="online"
          avatarUrl="https://api.adorable.io/avatars/240/thunder"
        />
        <Avatar size="xs" presence="offline" />
        <Avatar size="sm" presence="busy" />
        <Avatar size="lg" />
        <Avatar size="xl" />
        <Avatar size="xxl" />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
