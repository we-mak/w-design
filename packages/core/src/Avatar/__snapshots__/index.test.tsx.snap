// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar/> should renders correctly 1`] = `
<div>
  <figure
    class="sc-bdnxRM isQbNl"
    style="background-color: white;"
  >
    <span
      class="sc-dlnjwi kdLnmS"
    >
      Wm
    </span>
    <img
      alt="test"
      class="sc-hKFxyN jXkqiL sc-jSFjdj gZxmSH"
      data-src="https://api.adorable.io/avatars/240/thunder"
      height="44.8"
      loading="lazy"
      src="data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP88R8AAvUB+VkkrXoAAAAASUVORK5CYII="
      width="44.8"
    />
    <span
      class="sc-gtsrHT eGzKSC"
    />
  </figure>
  <figure
    class="sc-bdnxRM knbNFl"
    style="background-color: rgb(41, 55, 66);"
  >
    <span
      class="sc-dlnjwi bUaUPX"
    >
      Wm
    </span>
    <span
      class="sc-gtsrHT kKerWs"
    />
  </figure>
  <figure
    class="sc-bdnxRM iciBxR"
    style="background-color: rgb(57, 75, 89);"
  >
    <span
      class="sc-dlnjwi bGZMvM"
    >
      Wm
    </span>
    <span
      class="sc-gtsrHT fOmAO"
    />
  </figure>
  <figure
    class="sc-bdnxRM dzHwop"
    style="background-color: transparent;"
  >
    <span
      class="sc-dlnjwi kIeiXg"
    >
      Wm
    </span>
  </figure>
  <figure
    class="sc-bdnxRM beiRAp"
    style="background-color: rgb(116, 253, 127);"
  >
    <span
      class="sc-dlnjwi IsnqF"
    >
      Wm
    </span>
  </figure>
  <figure
    class="sc-bdnxRM hifa-dR"
    style="background-color: rgb(0, 154, 211);"
  >
    <span
      class="sc-dlnjwi CnYkr"
    >
      Wm
    </span>
  </figure>
</div>
`;
